transport-classes-io_uring/src/main/java/io/netty/channel/uring/IoUringBufferRing.java

```diff
@@ -0,0 +1,452 @@
+        //  https://github.com/axboe/liburing/blob/19134a8fffd406b22595a5813a3e319c19630ac9/src/include/liburing.h#L1561
+        long ioUringBufAddress = ioUringBufRingAddr + (long) Native.SIZEOF_IOURING_BUF * ringIndex;
+        PlatformDependent.putLong(ioUringBufAddress + Native.IOURING_BUFFER_OFFSETOF_ADDR, byteBuf.memoryAddress());
+        PlatformDependent.putInt(ioUringBufAddress + Native.IOURING_BUFFER_OFFSETOF_LEN, (short) byteBuf.capacity());
+        PlatformDependent.putShort(ioUringBufAddress + Native.IOURING_BUFFER_OFFSETOF_BID, bid);
+        if (needAdvance) {
+            advanceTail(1);
+        }
+    }
+
+    void appendBuffer(int count) {
+        int expectedIndex = nextIndex + count;
+        if (expectedIndex > entries) {
+            throw new IllegalStateException(
+                    String.format(
+                            "We want append %d buffer, but buffer ring is full. The ring hold %s buffers",
+                            count, entries)
+            );
+        }
+
+        for (int i = 0; i < count; i++) {
+            ByteBuf buffer = byteBufAllocator.ioBuffer(chunkSize);
+            short bid = nextIndex;
+            userspaceBufferHolder[bid] = buffer;
+            addToRing(nextIndex, false);
+            recycleBufferTasks[bid] = new Runnable() {
+                @Override
+                public void run() {
+                    addToRing(bid, true);
+                }
+            };
+            nextIndex++;
+        }
+        advanceTail(count);
+    }
+
+    ByteBuf borrowBuffer(int bid, int maxCap) {
+        ByteBuf byteBuf = userspaceBufferHolder[bid];
+        ByteBuf slice = byteBuf.retainedSlice(0, maxCap);
+        return new UserspaceIoUringBuffer(maxCap, (short) bid, slice);
+    }
+
+    private void advanceTail(int count) {
+        long tailFieldAddress = ioUringBufRingAddr + Native.IO_URING_BUFFER_RING_TAIL;
+        short oldTail = PlatformDependent.getShort(tailFieldAddress);
+        short newTail = (short) (oldTail + count);
+        PlatformDependent.putShortOrdered(tailFieldAddress, newTail);
+        hasSpareBuffer = true;
+    }
+
+    int entries() {
+        return entries;
+    }
+
+    short bufferGroupId() {
+        return bufferGroupId;
+    }
+
+    int chunkSize() {
+        return chunkSize;
+    }
+
+    boolean isFull() {
+        return nextIndex == entries;
+    }
+
+    long address() {
+        return ioUringBufRingAddr;
+    }
+
+    void close() {
+        Native.ioUringUnRegisterBufRing(ringFd, ioUringBufRingAddr, 4, 1);
+        for (ByteBuf byteBuf : userspaceBufferHolder) {
+            if (byteBuf != null) {
+                byteBuf.release();
+            }
+        }
+    }
+
+    class UserspaceIoUringBuffer extends AbstractReferenceCountedByteBuf {
+
+        private final short bid;
+
+        private final ByteBuf userspaceBuffer;
+
+        protected UserspaceIoUringBuffer(int maxCapacity, short bid, ByteBuf userspaceBuffer) {
```

Comment:
use `AbstractByteBuf unwrapped` and that would make possible, for each method, to use the `_` variant of methods

Reply:
The parameter userspaceBuffer comes from byteBuf.retainedSlice(0, maxCap). It seems that I can't obtain AbstractByteBuf. I didn't quite understand what you meant. Could you give me an example?

Reply:
I can have a look at this as a follow up